File: backend/app/api/endpoints/daily.py
"""
Daily report API endpoints.
"""
from datetime import date, datetime
from typing import Optional

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db
from app.services.daily_report_service import DailyReportService
from app.services.raw_sync_service import RawSyncService

router = APIRouter()


# ============ Request/Response Schemas ============

class DailyOverviewResponse(BaseModel):
    """Daily overview response."""
    summary: dict
    comparison: Optional[dict] = None
    daily: list
    start_date: str
    end_date: str


class TrendsResponse(BaseModel):
    """Trends response."""
    granularity: str
    data: list


class HourlyStatsResponse(BaseModel):
    """Hourly stats response."""
    date: str
    hourly: list


class TopKnowledgeBasesResponse(BaseModel):
    """Top knowledge bases response."""
    date: str
    items: list


class KnowledgeBaseListResponse(BaseModel):
    """Knowledge base list response."""
    items: list
    total: int
    page: int
    page_size: int


class KnowledgeBaseDetailResponse(BaseModel):
    """Knowledge base detail response."""
    id: int
    name: Optional[str] = None
    namespace: Optional[str] = None
    kb_type: Optional[str] = None
    is_active: Optional[bool] = None
    created_by_user_id: Optional[int] = None
    created_by_user_name: Optional[str] = None
    retrieval_config: Optional[dict] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class KnowledgeBaseStatsResponse(BaseModel):
    """Knowledge base stats response."""
    summary: dict
    daily: list


class QueryListResponse(BaseModel):
    """Query list response."""
    items: list
    total: int
    page: int
    page_size: int


class RagRecordDetailResponse(BaseModel):
    """RAG record detail response."""
    id: int
    raw_id: int
    knowledge_id: Optional[int] = None
    context_type: Optional[str] = None
    injection_mode: Optional[str] = None
    evaluation_status: Optional[str] = None
    evaluation_result_id: Optional[int] = None
    record_date: Optional[str] = None
    name: Optional[str] = None
    type_data: Optional[dict] = None
    extracted_text: Optional[str] = None
    created_at: Optional[str] = None


class SyncStatusResponse(BaseModel):
    """Sync status response."""
    raw_db_configured: bool
    hourly: Optional[dict] = None
    daily: Optional[dict] = None


class SyncTriggerRequest(BaseModel):
    """Sync trigger request."""
    sync_type: str = Field(default="hourly", description="Sync type: hourly, daily, or full")


class SyncTriggerResponse(BaseModel):
    """Sync trigger response."""
    status: str
    message: str
    result: Optional[dict] = None


# ============ Daily Overview Endpoints ============

@router.get("/overview", response_model=DailyOverviewResponse)
async def get_daily_overview(
    start_date: Optional[date] = Query(None, description="Start date (default: 7 days ago)"),
    end_date: Optional[date] = Query(None, description="End date (default: today)"),
    db: AsyncSession = Depends(get_db),
):
    """Get daily overview statistics."""
    service = DailyReportService(db)
    result = await service.get_daily_overview(start_date=start_date, end_date=end_date)
    return DailyOverviewResponse(**result)


@router.get("/trends", response_model=TrendsResponse)
async def get_trends(
    days: int = Query(7, ge=1, le=90, description="Number of days"),
    granularity: str = Query("day", description="Granularity: day or hour"),
    db: AsyncSession = Depends(get_db),
):
    """Get trend data over time."""
    if granularity not in ("day", "hour"):
        raise HTTPException(status_code=400, detail="Invalid granularity. Use 'day' or 'hour'")

    service = DailyReportService(db)
    result = await service.get_trends(days=days, granularity=granularity)
    return TrendsResponse(**result)


@router.get("/{target_date}/hourly", response_model=HourlyStatsResponse)
async def get_hourly_stats(
    target_date: date,
    db: AsyncSession = Depends(get_db),
):
    """Get hourly statistics for a specific date."""
    service = DailyReportService(db)
    hourly = await service.get_hourly_stats(target_date=target_date)
    return HourlyStatsResponse(date=target_date.isoformat(), hourly=hourly)


# ============ Knowledge Base Endpoints ============

@router.get("/knowledge-bases", response_model=KnowledgeBaseListResponse)
async def get_knowledge_bases(
    q: Optional[str] = Query(None, description="Search by id / creator / name / namespace"),
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
):
    """Get paginated global list of knowledge bases (from Raw DB).

    Results are sorted by recent 7-day usage (descending), then by id (descending).
    """
    service = DailyReportService(db)
    offset = (page - 1) * page_size
    items, total = await service.get_knowledge_base_list(
        q=q,
        limit=page_size,
        offset=offset,
    )
    return KnowledgeBaseListResponse(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
    )


@router.get("/knowledge-bases/top", response_model=TopKnowledgeBasesResponse)
async def get_top_knowledge_bases(
    target_date: Optional[date] = Query(None, description="Date for ranking (default: today)"),
    start_date: Optional[date] = Query(None, description="Start date for range ranking (optional)"),
    end_date: Optional[date] = Query(None, description="End date for range ranking (optional)"),
    limit: int = Query(10, ge=1, le=50, description="Number of results"),
    db: AsyncSession = Depends(get_db),
):
    """Get top knowledge bases by query count.

    - If start_date/end_date provided: return range ranking (sum across dates)
    - Else: return single-day ranking by target_date (default: today)
    """
    service = DailyReportService(db)
    items = await service.get_top_knowledge_bases(
        target_date=target_date,
        start_date=start_date,
        end_date=end_date,
        limit=limit,
    )

    # For response `date` field: prioritize explicit target_date; else show end_date/today
    resp_date = (target_date or end_date or date.today()).isoformat()

    return TopKnowledgeBasesResponse(
        date=resp_date,
        items=items,
    )


@router.get("/knowledge-bases/{kb_id}", response_model=KnowledgeBaseDetailResponse)
async def get_knowledge_base_detail(
    kb_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Get knowledge base detail from Raw DB."""
    service = DailyReportService(db)
    result = await service.get_knowledge_base_detail(kb_id)
    if not result:
        raise HTTPException(status_code=404, detail="Knowledge base not found")
    return KnowledgeBaseDetailResponse(**result)


@router.get("/knowledge-bases/{kb_id}/stats", response_model=KnowledgeBaseStatsResponse)
async def get_knowledge_base_stats(
    kb_id: int,
    days: int = Query(7, ge=1, le=90, description="Number of days"),
    db: AsyncSession = Depends(get_db),
):
    """Get statistics for a specific knowledge base."""
    service = DailyReportService(db)
    result = await service.get_knowledge_base_stats(kb_id=kb_id, days=days)
    return KnowledgeBaseStatsResponse(**result)


@router.get("/knowledge-bases/{kb_id}/queries", response_model=QueryListResponse)
async def get_knowledge_base_queries(
    kb_id: int,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    injection_mode: Optional[str] = Query(None, description="Filter by injection mode"),
    evaluation_status: Optional[str] = Query(None, description="Filter by evaluation status"),
    db: AsyncSession = Depends(get_db),
):
    """Get recent queries for a knowledge base."""
    service = DailyReportService(db)
    offset = (page - 1) * page_size
    items, total = await service.get_knowledge_base_queries(
        kb_id=kb_id,
        limit=page_size,
        offset=offset,
        injection_mode=injection_mode,
        evaluation_status=evaluation_status,
    )
    return QueryListResponse(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
    )


# ============ Global Query Endpoints ============

@router.get("/queries", response_model=QueryListResponse)
async def get_global_queries(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    injection_mode: Optional[str] = Query(None, description="Filter by injection mode (rag_retrieval, direct_injection, selected_documents)"),
    start_date: Optional[date] = Query(None, description="Filter by start date"),
    end_date: Optional[date] = Query(None, description="Filter by end date"),
    db: AsyncSession = Depends(get_db),
):
    """Get global query list (all queries across all knowledge bases).

    Returns queries with associated knowledge base information.
    """
    service = DailyReportService(db)
    offset = (page - 1) * page_size
    items, total = await service.get_global_queries(
        limit=page_size,
        offset=offset,
        injection_mode=injection_mode,
        start_date=start_date,
        end_date=end_date,
    )
    return QueryListResponse(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
    )


# ============ RAG Record Endpoints ============

@router.get("/rag-records/{record_id}", response_model=RagRecordDetailResponse)
async def get_rag_record_detail(
    record_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Get detailed RAG record."""
    service = DailyReportService(db)
    result = await service.get_rag_record_detail(record_id)
    if not result:
        raise HTTPException(status_code=404, detail="RAG record not found")
    return RagRecordDetailResponse(**result)


# ============ Sync Endpoints ============

@router.get("/sync/status", response_model=SyncStatusResponse)
async def get_raw_sync_status(
    db: AsyncSession = Depends(get_db),
):
    """Get raw data sync status."""
    service = RawSyncService(db)
    result = await service.get_sync_status()
    return SyncStatusResponse(**result)


@router.post("/sync/trigger", response_model=SyncTriggerResponse)
async def trigger_raw_sync(
    request: SyncTriggerRequest,
    background_tasks: BackgroundTasks,
    db: AsyncSession = Depends(get_db),
):
    """Manually trigger raw data sync."""
    service = RawSyncService(db)

    if request.sync_type == "hourly":
        result = await service.run_hourly_sync()
    elif request.sync_type == "daily":
        result = await service.run_daily_sync()
    elif request.sync_type == "full":
        # Run both hourly and daily
        hourly_result = await service.run_hourly_sync()
        daily_result = await service.run_daily_sync()
        result = {"hourly": hourly_result, "daily": daily_result}
    else:
        raise HTTPException(status_code=400, detail="Invalid sync_type. Use 'hourly', 'daily', or 'full'")

    return SyncTriggerResponse(
        status=result.get("status", "unknown"),
        message="Sync completed",
        result=result,
    )

File: frontend/src/apis/daily.ts
/**
 * Daily report API client
 */

import { apiUrl } from './base'

export interface DailySummary {
  total_queries: number
  rag_retrieval_count: number
  direct_injection_count: number
  selected_documents_count: number
  active_kb_count: number
  active_user_count: number
}

export interface DailyComparison {
  total_queries_change: number
  rag_retrieval_change: number
}

export interface DailyDataPoint {
  date: string
  total_queries: number
  rag_retrieval_count: number
  direct_injection_count: number
  selected_documents_count: number
  active_kb_count?: number
  active_user_count?: number
}

export interface DailyOverviewResponse {
  summary: DailySummary
  comparison: DailyComparison | null
  daily: DailyDataPoint[]
  start_date: string
  end_date: string
}

export interface HourlyDataPoint {
  hour: number
  total_queries: number
  rag_retrieval_count: number
  direct_injection_count: number
  selected_documents_count: number
}

export interface TrendsResponse {
  granularity: 'day' | 'hour'
  data: DailyDataPoint[]
}

export interface KnowledgeBaseItem {
  knowledge_id: number
  knowledge_name: string | null
  namespace: string | null
  created_by_user_id?: number | null
  created_by_user_name?: string | null
  description?: string | null
  kb_type?: string | null
  created_at?: string | null
  updated_at?: string | null
  recent_7d_queries?: number
  recent_7d_used?: boolean
  // 注意：全局列表页当前不展示统计列，这些字段在列表接口中可能为 0
  total_queries: number
  rag_retrieval_count: number
  direct_injection_count: number
  selected_documents_count: number
  rank?: number
  primary_mode?: string
}

export interface KnowledgeBaseDetail {
  id: number
  name: string | null
  namespace: string | null
  kb_type: string | null
  is_active: boolean | null
  created_by_user_id?: number | null
  created_by_user_name?: string | null
  retrieval_config: {
    retriever_name: string | null
    retrieval_mode: string | null
    top_k: number | null
    score_threshold: number | null
    embedding_model: string | null
  } | null
  created_at: string | null
  updated_at: string | null
}

export interface KnowledgeBaseStats {
  summary: {
    knowledge_id: number
    knowledge_name: string | null
    namespace: string | null
    total_queries: number
    rag_retrieval_count: number
    direct_injection_count: number
    selected_documents_count: number
  }
  daily: DailyDataPoint[]
}

export interface QueryItem {
  id: number
  raw_id: number
  record_date: string | null
  context_type: string | null
  injection_mode: string | null
  evaluation_status: string | null
  query: string | null
  chunks_count: number | null
  sources: Array<{ index: number; kb_id: number; title: string }> | null
  created_at: string | null
}

export interface GlobalQueryItem extends QueryItem {
  knowledge_id: number | null
  knowledge_name: string | null
  namespace: string | null
}

export interface RagRecordDetail {
  id: number
  raw_id: number
  knowledge_id: number | null
  context_type: string | null
  injection_mode: string | null
  evaluation_status: string | null
  evaluation_result_id: number | null
  record_date: string | null
  name: string | null
  type_data: Record<string, unknown> | null
  extracted_text: string | null
  created_at: string | null
}

export interface SyncStatus {
  raw_db_configured: boolean
  hourly: {
    last_sync_time: string | null
    last_raw_id: number
    status: string | null
    records_synced: number
    error_message: string | null
  } | null
  daily: {
    last_sync_time: string | null
    status: string | null
    error_message: string | null
  } | null
}

export interface PaginatedResponse<T> {
  items: T[]
  total: number
  page: number
  page_size: number
}

/**
 * Get daily overview statistics
 */
export async function getDailyOverview(params?: {
  start_date?: string
  end_date?: string
}): Promise<DailyOverviewResponse> {
  const searchParams = new URLSearchParams()
  if (params?.start_date) searchParams.set('start_date', params.start_date)
  if (params?.end_date) searchParams.set('end_date', params.end_date)

  const url = apiUrl(`/daily/overview${searchParams.toString() ? `?${searchParams}` : ''}`)
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to fetch daily overview')
  return response.json()
}

/**
 * Get trend data
 */
export async function getDailyTrends(params?: {
  days?: number
  granularity?: 'day' | 'hour'
}): Promise<TrendsResponse> {
  const searchParams = new URLSearchParams()
  if (params?.days) searchParams.set('days', params.days.toString())
  if (params?.granularity) searchParams.set('granularity', params.granularity)

  const url = apiUrl(`/daily/trends${searchParams.toString() ? `?${searchParams}` : ''}`)
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to fetch trends')
  return response.json()
}

/**
 * Get hourly stats for a specific date
 */
export async function getHourlyStats(date: string): Promise<{
  date: string
  hourly: HourlyDataPoint[]
}> {
  const url = apiUrl(`/daily/${date}/hourly`)
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to fetch hourly stats')
  return response.json()
}

/**
 * Get top knowledge bases
 */
export async function getTopKnowledgeBases(params?: {
  target_date?: string
  start_date?: string
  end_date?: string
  limit?: number
}): Promise<{ date: string; items: KnowledgeBaseItem[] }> {
  const searchParams = new URLSearchParams()
  if (params?.target_date) searchParams.set('target_date', params.target_date)
  if (params?.start_date) searchParams.set('start_date', params.start_date)
  if (params?.end_date) searchParams.set('end_date', params.end_date)
  if (params?.limit) searchParams.set('limit', params.limit.toString())

  const url = apiUrl(
    `/daily/knowledge-bases/top${searchParams.toString() ? `?${searchParams}` : ''}`
  )
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to fetch top knowledge bases')
  return response.json()
}

/**
 * Get knowledge base list
 */
export async function getKnowledgeBases(params?: {
  q?: string
  page?: number
  page_size?: number
}): Promise<PaginatedResponse<KnowledgeBaseItem>> {
  const searchParams = new URLSearchParams()
  if (params?.q) searchParams.set('q', params.q)
  if (params?.page) searchParams.set('page', params.page.toString())
  if (params?.page_size) searchParams.set('page_size', params.page_size.toString())

  const url = apiUrl(
    `/daily/knowledge-bases${searchParams.toString() ? `?${searchParams}` : ''}`
  )
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to fetch knowledge bases')
  return response.json()
}

/**
 * Get knowledge base detail
 */
export async function getKnowledgeBaseDetail(kbId: number): Promise<KnowledgeBaseDetail> {
  const url = apiUrl(`/daily/knowledge-bases/${kbId}`)
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to fetch knowledge base detail')
  return response.json()
}

/**
 * Get knowledge base stats
 */
export async function getKnowledgeBaseStats(
  kbId: number,
  params?: { days?: number }
): Promise<KnowledgeBaseStats> {
  const searchParams = new URLSearchParams()
  if (params?.days) searchParams.set('days', params.days.toString())

  const url = apiUrl(
    `/daily/knowledge-bases/${kbId}/stats${searchParams.toString() ? `?${searchParams}` : ''}`
  )
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to fetch knowledge base stats')
  return response.json()
}

/**
 * Get knowledge base queries
 */
export async function getKnowledgeBaseQueries(
  kbId: number,
  params?: {
    page?: number
    page_size?: number
    injection_mode?: string
    evaluation_status?: string
  }
): Promise<PaginatedResponse<QueryItem>> {
  const searchParams = new URLSearchParams()
  if (params?.page) searchParams.set('page', params.page.toString())
  if (params?.page_size) searchParams.set('page_size', params.page_size.toString())
  if (params?.injection_mode) searchParams.set('injection_mode', params.injection_mode)
  if (params?.evaluation_status) searchParams.set('evaluation_status', params.evaluation_status)

  const url = apiUrl(
    `/daily/knowledge-bases/${kbId}/queries${searchParams.toString() ? `?${searchParams}` : ''}`
  )
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to fetch queries')
  return response.json()
}

/**
 * Get RAG record detail
 */
export async function getRagRecordDetail(recordId: number): Promise<RagRecordDetail> {
  const url = apiUrl(`/daily/rag-records/${recordId}`)
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to fetch RAG record detail')
  return response.json()
}

/**
 * Get sync status
 */
export async function getSyncStatus(): Promise<SyncStatus> {
  const url = apiUrl('/daily/sync/status')
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to fetch sync status')
  return response.json()
}

/**
 * Trigger sync
 */
export async function triggerSync(syncType: 'hourly' | 'daily' | 'full'): Promise<{
  status: string
  message: string
  result: Record<string, unknown> | null
}> {
  const url = apiUrl('/daily/sync/trigger')
  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ sync_type: syncType }),
  })
  if (!response.ok) throw new Error('Failed to trigger sync')
  return response.json()
}

/**
 * Get global queries (all queries across all knowledge bases)
 */
export async function getGlobalQueries(params?: {
  page?: number
  page_size?: number
  injection_mode?: string
  start_date?: string
  end_date?: string
}): Promise<PaginatedResponse<GlobalQueryItem>> {
  const searchParams = new URLSearchParams()
  if (params?.page) searchParams.set('page', params.page.toString())
  if (params?.page_size) searchParams.set('page_size', params.page_size.toString())
  if (params?.injection_mode) searchParams.set('injection_mode', params.injection_mode)
  if (params?.start_date) searchParams.set('start_date', params.start_date)
  if (params?.end_date) searchParams.set('end_date', params.end_date)

  const url = apiUrl(
    `/daily/queries${searchParams.toString() ? `?${searchParams}` : ''}`
  )
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to fetch global queries')
  return response.json()
}
